Enable custom tiki Parser

jevertz opened the pull request:
This makes it possible to use a custom Tika Parser.
To use the parser, just an addition to _settings.json is necessary.
Here's an example
    "custom_tika_parsers": [
      {
        "class_name": "com.monotype.OTParser",
        "path_to_jar": "/Users/jevertz/workspace/testfscrawler/svn-search/customTikaParsers/target/otparser-0.0.1-SNAPSHOT.jar",
        "mime_types": ["application/x-font-ttf", "application/x-font-otf"]
      }
    ]

The jar file does not need to be on the class path. 
The parser needs to be a Tika parser, like described [here](https://tika.apache.org/1.17/parser_guide.html)

dadoonet commented:
Very interesting. I like the idea.

I wonder if it's absolutely needed to provide the `tika-config.xml` file BTW.

I'll give a proper review later but thanks for sharing your code with this PR!

jevertz commented:
You’re. welcome.
Actually I implemented first passing a Tika Config which can be useful, but not for my use-case.
So I took a different approach and removed the need for a Tika Config file later.

Best

Jochen Evertz

Von meinem iPhone gesendet



Jochen Evertz
Software Engineer
Monotype GmbH
Bergmannstr. 102
10961 Berlin
Germany
Phone:  +49 (30) 69596 221
Mobile: +49 (151) 17149034
                              
                              
monotype.com
Geschäftsführer / Managing Director: Christopher Kollat
Sitz Bad Homburg, HRB 10375, Amtsgericht Bad Homburg / Commercial Register of the Bad Homburg Local Court, HRB 10375

This e-mail message and any accompanying documentation is intended for the addressee only and may contain legally privileged and confidential information. If you are not the addressee any use of this information is strictly prohibited. If you have received this e-mail in error please notify us immediately by reply e-mail or telephone call on +49 (6172) 484-0 and delete this e-mail from your systems.

Am 02.02.2018 um 18:35 schrieb David Pilato <notifications@github.com<mailto:notifications@github.com>>:


Very interesting. I like the idea.

I wonder if it's absolutely needed to provide the tika-config.xml file BTW.

I'll give a proper review later but thanks for sharing your code with this PR!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23issuecomment-2D362644819&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=a-v_BHjTtHM3oQdR3R2WZpIf6JufkkLLKjogymRRzA4&s=RkkFis8gn21WJYEddt5E_czRkygrpoh7ZU4KycoUaLk&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AUX8wsSsKKBYTfvNPpmX7SQy6poQtYilks5tQ0C5gaJpZM4R3S2m&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=a-v_BHjTtHM3oQdR3R2WZpIf6JufkkLLKjogymRRzA4&s=qiVQmwIIJyQXo3DJAtc5CNLRuZGxGvWzYK4GdM5bKWc&e=>.

dadoonet commented:
Hi @jevertz 

Would you mind rebasing your code on latest master branch?

dadoonet commented:
Also I think it would be good to document that change in the README.

jevertz commented:
Hi David,

might take a little since the ES version changed. Don’t want to break my setup.
But not weeks.

Best

Jochen

Von: David Pilato <notifications@github.com>
Antworten an: dadoonet/fscrawler <reply@reply.github.com>
Datum: Donnerstag, 15. Februar 2018 um 23:48
An: dadoonet/fscrawler <fscrawler@noreply.github.com>
Cc: "Evertz, Jochen" <Jochen.Evertz@monotype.com>, Mention <mention@noreply.github.com>
Betreff: Re: [dadoonet/fscrawler] Enable custom tiki Parser (#498)


Hi @jevertz<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_jevertz&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=xmumm6BweYLfjE07JW_nZ671Ou2JaCph79gQfldssrc&s=E8j8bMdnDV_bB6ID8P7HbfLHLKetX2MP96bSt8M3eDw&e=>

Would you mind rebasing your code on latest master branch?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23issuecomment-2D366087903&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=xmumm6BweYLfjE07JW_nZ671Ou2JaCph79gQfldssrc&s=DRA507VIIyWX3sJzKbqbtVkWgrtYMv8xr-CNc2t2XsY&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AUX8wkJh8j8hL-2DIMnGE4PsjpdsSsHV-2Dpks5tVLQjgaJpZM4R3S2m&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=xmumm6BweYLfjE07JW_nZ671Ou2JaCph79gQfldssrc&s=lJRAfFhcLeYBljJdCub_GBRv3nS-juy3zCISDE46QOA&e=>.


Jochen Evertz
Software Engineer
Monotype GmbH
Bergmannstr. 102
10961 Berlin
Germany
Phone:  +49 (30) 69596 221
Mobile: +49 (151) 17149034
                              
                              
monotype.com
Geschäftsführer / Managing Director: Christopher Kollat
Sitz Bad Homburg, HRB 10375, Amtsgericht Bad Homburg / Commercial Register of the Bad Homburg Local Court, HRB 10375

This e-mail message and any accompanying documentation is intended for the addressee only and may contain legally privileged and confidential information. If you are not the addressee any use of this information is strictly prohibited. If you have received this e-mail in error please notify us immediately by reply e-mail or telephone call on +49 (6172) 484-0 and delete this e-mail from your systems.

dadoonet commented on README.md:
```diff
@@ -1183,6 +1184,181 @@ to compute the checksum, such as `MD5` or `SHA-1`.
+It might occur that one or more existing Tika parsers do not provide the intended information, or just do not exist.
+This setting allows to use a custom parser instead.
+The parsers must be provided as a .jar, but does not need to be on any classpath.
+Note that this is an array. Here an example for just one
```
NIT: 

> Here an example for just one

to

> Here an example for just one:

dadoonet commented on README.md:
```diff
@@ -1183,6 +1184,181 @@ to compute the checksum, such as `MD5` or `SHA-1`.
+
+Some info about creating a custom parser is available [here](https://tika.apache.org/1.17/parser_guide.html)
+Or use a existing parser as a blueprint. Make sure to choose the correct branch.
+At the time of this writing fscrawler uses Tika 1.17, while on github the main branch is 2.x.
```
> the main branch is 2.x

to 

> the main Tika branch is 2.x

dadoonet commented on README.md:
```diff
@@ -1183,6 +1184,181 @@ to compute the checksum, such as `MD5` or `SHA-1`.
+The parsers from ["branch_1x"](https://github.com/apache/tika/tree/branch_1x/tika-parsers/src/main/java/org/apache/tika/parser) should work fine.
+
+To build the custom parser separately, a pom file can be derived from the tika-parsers [pom.xml](https://github.com/apache/tika/blob/branch_1x/tika-parsers/pom.xml).
+Probably a lot can be left out. Here is an example which required fontbox (guess still to long, but worked).
```
WDYM with "guess still to long, but worked"

too long? Is it a comment you are making here? If so I believe this is not needed.

jevertz replied:
Just to indicate that I didn’t really evaluated the Pom.xml, but stopped when it build without errors.
Should get marked someway as to-do

jevertz replied:
Changed it to
(The exclusions are copied 1:1 from fscrawler's pom.xml, to be on the safe side)

dadoonet commented on README.md:
```diff
@@ -1183,6 +1184,181 @@ to compute the checksum, such as `MD5` or `SHA-1`.
+                    <groupId>com.uwyn</groupId>
+                    <artifactId>jhighlight</artifactId>
+                </exclusion>
+                <!-- ES core already has these -->
```
ES Core -> FSCrawler Core ?

jevertz replied:
I copied the excludes from fscrawler's root pom.xml. The line is still there at the time of this writing. Can it be removed, perhaps?

dadoonet replied:
That's true. Just remove it here but I will also remove this in the future from the main `pom.xml` indeed.

dadoonet commented on README.md:
```diff
@@ -1183,6 +1184,181 @@ to compute the checksum, such as `MD5` or `SHA-1`.
+    <version>0.0.1-SNAPSHOT</version>
+
+    <properties>
+        <maven.compiler.source>1.7</maven.compiler.source>
```
Not an issue but I'd recommend using 1.8 as the version. I know that 1.7 will work but this JVM is no longer supported so no need to advertise it :)

jevertz replied:
You’re completely right. I just didn’t notice the setting.
I’ll change it to 1.8

dadoonet commented on tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java:
```diff
@@ -78,6 +87,8 @@ private static void initParser(Fs fs) {
             PDFParser pdfParser = new PDFParser();
             DefaultParser defaultParser;
 
+
```
NIT: Not needed "change". :)

dadoonet commented on tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java:
```diff
@@ -94,11 +105,62 @@ private static void initParser(Fs fs) {
+                Integer counter = 0;
+                Parser PARSERS[] = new Parser[fs.getCustomTikaParsers().size()+2];
+
+
```
Can you remove some blank lines here? One is enough.

jevertz replied:
Sure. Was not intentional.

dadoonet commented on tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java:
```diff
@@ -94,11 +105,62 @@ private static void initParser(Fs fs) {
+                // to collect all Mediatypes handled by custom parser to exclude them form the DefaultParser
+                List<MediaType> excludeMediaTypes = new ArrayList<MediaType>();
+
+
```
Same here

dadoonet commented on settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/Fs.java:
```diff
@@ -212,10 +214,28 @@ public Builder setOcr(Ocr ocr) {
+        }
+
+        public Builder addTikaCustomParsers(CustomTikaParser customTikaParser) {
+            if (this.customTikaParsers == null) {
```
I believe this can not normally happen?

dadoonet commented on settings/src/test/java/fr/pilato/elasticsearch/crawler/fs/settings/FsSettingsParserTest.java:
```diff
@@ -53,6 +54,7 @@
             .setUpdateRate(TimeValue.timeValueMinutes(5))
             .setIndexContent(true)
             .setOcr(OCR_FULL)
+            .setTikaCustomParsers(new ArrayList<>())
```
I think you should add at least a singleton List of one `CustomTikaParser` element here to make sure serialisation and deserialisation is working well.

dadoonet commented on tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java:
```diff
@@ -94,11 +105,62 @@ private static void initParser(Fs fs) {
+                        Parser customParserDecorated = ParserDecorator.withTypes(customParser, customMediaTypes);
+                        PARSERS[counter] = customParserDecorated;
+
+
```
Same.

dadoonet commented on tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java:
```diff
@@ -94,11 +105,62 @@ private static void initParser(Fs fs) {
+
+
+
+                    } catch (IOException e) {
```
I think you can just do here something like:

```java
} catch (IOException|ClassNotFoundException|InstantiationException|IllegalAccessException e) {
    logger.error("Caught {}: {}", e.getClass().getSimpleName(), e.getMessage());
}
```

Or just catch any `Exception`?

dadoonet commented on tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java:
```diff
@@ -94,11 +105,62 @@ private static void initParser(Fs fs) {
+                        logger.error("Caught InstantiationException:" + e.getMessage());
+                    } catch (IllegalAccessException e) {
+                        logger.error("Caught IllegalAccessException:" + e.getMessage());
+                    }/*catch (TikaException te) {
```
Probably we don't need that?

jevertz commented:
You’re welcome!
Sorry for the delay. I‘ll merge your corrections ASAP - what probably means next Monday.
I wasn‘t sure whether it’s a good idea to include it in a full example since an error will be thrown if the custom jar doesn’t exist.
Or do you think it is useful to add a complete custom parser?
I didn’t do it since I thought it might confuse most users.

Best

Jochen

Von meinem iPhone gesendet



Jochen Evertz
Software Engineer
Monotype GmbH
Bergmannstr. 102
10961 Berlin
Germany
Phone:  +49 (30) 69596 221
Mobile: +49 (151) 17149034
                              
                              
monotype.com
Geschäftsführer / Managing Director: Christopher Kollat
Sitz Bad Homburg, HRB 10375, Amtsgericht Bad Homburg / Commercial Register of the Bad Homburg Local Court, HRB 10375

This e-mail message and any accompanying documentation is intended for the addressee only and may contain legally privileged and confidential information. If you are not the addressee any use of this information is strictly prohibited. If you have received this e-mail in error please notify us immediately by reply e-mail or telephone call on +49 (6172) 484-0 and delete this e-mail from your systems.

Am 22.02.2018 um 17:57 schrieb David Pilato <notifications@github.com<mailto:notifications@github.com>>:


@dadoonet requested changes on this pull request.

That looks very promising. Thanks!

I added a lot of "format" comments.
In the README, I'm always trying to print a full setting configuration in https://github.com/dadoonet/fscrawler#job-file-specification<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler-23job-2Dfile-2Dspecification&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=CbiAJeT8nYQ14c5zldlmot9i6_O6egGQ-rXLUIc5OcE&e=>
May be you could add your changes as well here. IIRC I'm using part of the output of the FsSettingsParserTest execution.

________________________________

In README.md<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170013425&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=9Ocn4Vo62GlzO_Ql5FtgaTCVik8Pek50zD0m6f8LfoU&e=>:

> @@ -1183,6 +1184,181 @@ to compute the checksum, such as `MD5` or `SHA-1`.
 }
 ```

+#### Custom Tika Parsers
+
+It might occur that one or more existing Tika parsers do not provide the intended information, or just do not exist.
+This setting allows to use a custom parser instead.
+The parsers must be provided as a .jar, but does not need to be on any classpath.
+Note that this is an array. Here an example for just one


NIT:

Here an example for just one

to

Here an example for just one:

________________________________

In README.md<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170013848&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=CxBc6ZOv6_w2sHBZ3nJ75TALiu2lBqwv3nY5-CggWwQ&e=>:

> +  "name": "test",
+  "fs": {
+    "custom_tika_parsers": [
+      {
+        "class_name": "org.me.MyParser",
+        "path_to_jar": "/some/full/path/to/myParser-0.0.1-SNAPSHOT.jar",
+        "mime_types": ["application/dns", "or-another-mimetype-from-tika"]
+      }
+    ]
+  }
+}
+```
+
+Some info about creating a custom parser is available [here](https://tika.apache.org/1.17/parser_guide.html)
+Or use a existing parser as a blueprint. Make sure to choose the correct branch.
+At the time of this writing fscrawler uses Tika 1.17, while on github the main branch is 2.x.


the main branch is 2.x

to

the main Tika branch is 2.x

________________________________

In README.md<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170014423&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=U2ocKd9u440aZzJ5fBUF_IqIaO_ZTsi2_greirbMCKg&e=>:

> +        "class_name": "org.me.MyParser",
+        "path_to_jar": "/some/full/path/to/myParser-0.0.1-SNAPSHOT.jar",
+        "mime_types": ["application/dns", "or-another-mimetype-from-tika"]
+      }
+    ]
+  }
+}
+```
+
+Some info about creating a custom parser is available [here](https://tika.apache.org/1.17/parser_guide.html)
+Or use a existing parser as a blueprint. Make sure to choose the correct branch.
+At the time of this writing fscrawler uses Tika 1.17, while on github the main branch is 2.x.
+The parsers from ["branch_1x"](https://github.com/apache/tika/tree/branch_1x/tika-parsers/src/main/java/org/apache/tika/parser) should work fine.
+
+To build the custom parser separately, a pom file can be derived from the tika-parsers [pom.xml](https://github.com/apache/tika/blob/branch_1x/tika-parsers/pom.xml).
+Probably a lot can be left out. Here is an example which required fontbox (guess still to long, but worked).


WDYM with "guess still to long, but worked"

too long? Is it a comment you are making here? If so I believe this is not needed.

________________________________

In README.md<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170014786&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=_oZ4mv7eqPlP6_mk8rpBZGNEttJsy8aA_3c3oSJo3nM&e=>:

> +                <!-- Not Apache2 License compatible -->
+                <exclusion>
+                    <groupId>edu.ucar</groupId>
+                    <artifactId>grib</artifactId>
+                </exclusion>
+                <!-- Not Apache2 License compatible -->
+                <exclusion>
+                    <groupId>edu.ucar</groupId>
+                    <artifactId>netcdf4</artifactId>
+                </exclusion>
+                <!-- Not Apache2 License compatible -->
+                <exclusion>
+                    <groupId>com.uwyn</groupId>
+                    <artifactId>jhighlight</artifactId>
+                </exclusion>
+                <!-- ES core already has these -->


ES Core -> FSCrawler Core ?

________________________________

In README.md<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170015211&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=bORjvLWce7otmM17ZDrzEtvaKStY3BInH99hkzg0-MQ&e=>:

> +Probably a lot can be left out. Here is an example which required fontbox (guess still to long, but worked).
+
+<details><summary>Example pom.xml</summary>
+<p>
+
+
+```
+<project xmlns="http://maven.apache.org/POM/4.0.0">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.me<http://org.me></groupId>
+    <artifactId>myParser</artifactId>
+    <version>0.0.1-SNAPSHOT</version>
+
+    <properties>
+        <maven.compiler.source>1.7</maven.compiler.source>


Not an issue but I'd recommend using 1.8 as the version. I know that 1.7 will work but this JVM is no longer supported so no need to advertise it :)

________________________________

In tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170016636&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=MHKjYAfcmjSVM5RF5zcn74T87tiqX5LX9G1JXKns80M&e=>:

> @@ -78,6 +87,8 @@ private static void initParser(Fs fs) {
             PDFParser pdfParser = new PDFParser();
             DefaultParser defaultParser;

+


NIT: Not needed "change". :)

________________________________

In tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170016758&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=-FNrL-l1u1cm4T_d8Oyy6yOhHyJ-V5N_cGSpnA-ViRM&e=>:

>
-            parser = new AutoDetectParser(PARSERS);
+            // load custom parsers if defined in config
+            if (fs.getCustomTikaParsers().size() > 0) {
+                Integer counter = 0;
+                Parser PARSERS[] = new Parser[fs.getCustomTikaParsers().size()+2];
+
+


Can you remove some blank lines here? One is enough.

________________________________

In tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170016800&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=-Z0iXtbo362zK486_ksJrqYNqrAVU1WMvf9_1eCAMb8&e=>:

>
-            parser = new AutoDetectParser(PARSERS);
+            // load custom parsers if defined in config
+            if (fs.getCustomTikaParsers().size() > 0) {
+                Integer counter = 0;
+                Parser PARSERS[] = new Parser[fs.getCustomTikaParsers().size()+2];
+
+
+
+                // to collect all Mediatypes handled by custom parser to exclude them form the DefaultParser
+                List<MediaType> excludeMediaTypes = new ArrayList<MediaType>();
+
+


Same here

________________________________

In settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/Fs.java<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170017268&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=FEskqfbE6p0FZo8-wlHHmgbe9BuB3CSSaciJTn9ojjk&e=>:

> @@ -212,10 +214,28 @@ public Builder setOcr(Ocr ocr) {
             return this;
         }

+        public Builder setTikaCustomParsers(List<CustomTikaParser> customTikaParsers) {
+            this.customTikaParsers = customTikaParsers;
+            return this;
+        }
+
+        public Builder addTikaCustomParsers(CustomTikaParser customTikaParser) {
+            if (this.customTikaParsers == null) {


I believe this can not normally happen?

________________________________

In settings/src/test/java/fr/pilato/elasticsearch/crawler/fs/settings/FsSettingsParserTest.java<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170017419&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=Uq_u2yPneiqxWkTncCZzQVHU3yVpBr9rGeob6csQWY8&e=>:

> @@ -53,6 +54,7 @@
             .setUpdateRate(TimeValue.timeValueMinutes(5))
             .setIndexContent(true)
             .setOcr(OCR_FULL)
+            .setTikaCustomParsers(new ArrayList<>())


I think you should add at least a singleton List of one CustomTikaParser element here to make sure serialisation and deserialisation is working well.

________________________________

In tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170017516&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=9WlUVzG73WGSQkofHL6s7eKsRLRsgL_G7K-xODTeq-8&e=>:

> +
+
+                for (CustomTikaParser customTikaParser : fs.getCustomTikaParsers()) {
+                    counter += 1;
+                    try {
+                        URL[] jarUrl = { new URL("jar:file:" + customTikaParser.getPathToJar()+"!/") };
+                        URLClassLoader urlClassLoader = URLClassLoader.newInstance(jarUrl);
+                        Class customParserClass = urlClassLoader.loadClass(customTikaParser.getClassName());
+                        Parser customParser = (Parser) customParserClass.newInstance();
+                        String[] customMimeStrings = new String[customTikaParser.getMimeTypes().size()];
+                        Set<MediaType> customMediaTypes = MediaType.set(customTikaParser.getMimeTypes().toArray(customMimeStrings));
+                        excludeMediaTypes.addAll(customMediaTypes);
+                        Parser customParserDecorated = ParserDecorator.withTypes(customParser, customMediaTypes);
+                        PARSERS[counter] = customParserDecorated;
+
+


Same.

________________________________

In tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170018299&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=gr69M29dRVbfcGw8zEwZX7JRcjy9Al8wW7xixtnnxnk&e=>:

> +                for (CustomTikaParser customTikaParser : fs.getCustomTikaParsers()) {
+                    counter += 1;
+                    try {
+                        URL[] jarUrl = { new URL("jar:file:" + customTikaParser.getPathToJar()+"!/") };
+                        URLClassLoader urlClassLoader = URLClassLoader.newInstance(jarUrl);
+                        Class customParserClass = urlClassLoader.loadClass(customTikaParser.getClassName());
+                        Parser customParser = (Parser) customParserClass.newInstance();
+                        String[] customMimeStrings = new String[customTikaParser.getMimeTypes().size()];
+                        Set<MediaType> customMediaTypes = MediaType.set(customTikaParser.getMimeTypes().toArray(customMimeStrings));
+                        excludeMediaTypes.addAll(customMediaTypes);
+                        Parser customParserDecorated = ParserDecorator.withTypes(customParser, customMediaTypes);
+                        PARSERS[counter] = customParserDecorated;
+
+
+
+                    } catch (IOException e) {


I think you can just do here something like:

} catch (IOException|ClassNotFoundException|InstantiationException|IllegalAccessException e) {
    logger.error("Caught {}: {}", e.getClass().getSimpleName(), e.getMessage());
}

Or just catch any Exception?

________________________________

In tika/src/main/java/fr/pilato/elasticsearch/crawler/fs/tika/TikaInstance.java<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23discussion-5Fr170018611&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=sia4NPm_6ksNLRyIpq6I8QMjCDDTyLlZJL1xdsH7iTM&e=>:

> +                        Set<MediaType> customMediaTypes = MediaType.set(customTikaParser.getMimeTypes().toArray(customMimeStrings));
+                        excludeMediaTypes.addAll(customMediaTypes);
+                        Parser customParserDecorated = ParserDecorator.withTypes(customParser, customMediaTypes);
+                        PARSERS[counter] = customParserDecorated;
+
+
+
+                    } catch (IOException e) {
+                        logger.error("Caught IOException:" + e.getMessage());
+                    } catch (ClassNotFoundException e) {
+                        logger.error("Caught ClassNotFoundException:" + e.getMessage());
+                    } catch (InstantiationException e) {
+                        logger.error("Caught InstantiationException:" + e.getMessage());
+                    } catch (IllegalAccessException e) {
+                        logger.error("Caught IllegalAccessException:" + e.getMessage());
+                    }/*catch (TikaException te) {


Probably we don't need that?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23pullrequestreview-2D98633158&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=5CEW9kreeqp8ElYFLzc-IyaaFZO5bEx6tW3HBY0VPHw&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AUX8wlRkvmkB5I2ShSPGZuXRhEAQapJaks5tXZxXgaJpZM4R3S2m&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=tdsDdwlAkvd1kVgEnB_G-7vxIy50RwUc_El2XpqR0Ts&s=AfxxklTQhPIwnN-jfQ88PrRspwWefFC6l4r_RmiOI7Q&e=>.

dadoonet commented:
Ideally you would create a submodule which is added as a test dependency in the integration tests module.
So we have something that compiles and that is tested.

Then link in the README to the module and the test class.

WDYT?

jevertz commented:
OK, I‘ll see to it.
Might be clearer to use an example with a custom extension instead of mime-type.
Have to look into that. Do you perhaps know a starting point to do that when setting up a Tika parser from FSCrawler?

Von meinem iPhone gesendet



Jochen Evertz
Software Engineer
Monotype GmbH
Bergmannstr. 102
10961 Berlin
Germany
Phone:  +49 (30) 69596 221
Mobile: +49 (151) 17149034
                              
                              
monotype.com
Geschäftsführer / Managing Director: Christopher Kollat
Sitz Bad Homburg, HRB 10375, Amtsgericht Bad Homburg / Commercial Register of the Bad Homburg Local Court, HRB 10375

This e-mail message and any accompanying documentation is intended for the addressee only and may contain legally privileged and confidential information. If you are not the addressee any use of this information is strictly prohibited. If you have received this e-mail in error please notify us immediately by reply e-mail or telephone call on +49 (6172) 484-0 and delete this e-mail from your systems.

Am 23.02.2018 um 22:12 schrieb David Pilato <notifications@github.com<mailto:notifications@github.com>>:


Ideally you would create a submodule which is added as a test dependency in the integration tests module.
So we have something that compiles and that is tested.

Then link in the README to the module and the test class.

WDYT?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23issuecomment-2D368139925&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=ZTIgMwMW6ShgsohC7ThPAheiQpp_ers2S1ogRvnGxf0&s=QE2Ol8bkmZPhYbdDgzAJFyogog9iq3CMtgRDdDTRLgU&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AUX8wqz5I61hvqeWfUJSxmPyWtNAwbywks5tXym9gaJpZM4R3S2m&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=ZTIgMwMW6ShgsohC7ThPAheiQpp_ers2S1ogRvnGxf0&s=ibuS_7mOfxwotLiR4LfOrW8A_cK0V4TvRN_2b_xKKPk&e=>.

dadoonet commented:
> Do you perhaps know a starting point to do that when setting up a Tika parser from FSCrawler?

I'm confused. Isn't what you proposed with that:

```js
{
  "name": "test",
  "fs": {
    "custom_tika_parsers": [
      {
        "class_name": "org.me.MyParser",
        "path_to_jar": "/some/full/path/to/myParser-0.0.1-SNAPSHOT.jar",
        "mime_types": ["application/dns", "or-another-mimetype-from-tika"]
      }
    ]
  }
}
```

I mean that you could create `fscrawler-sample-parser.jar` which you can add in a test?
If you don't know how to do it, no worries, I can do it in a follow up PR.

jevertz commented:
Sorry, for causing this confusion. All my fault. That was expressed bad from me.
At the moment I use the mime_types. However there are files which do not have a mime type, but a custom extension.
That means I need to pass a custom mime type to Tika.
Thought you might have something like that before, so I dared to asked.

I’ll just see how to accomplish it. The documentation of Tika<https://tika.apache.org/1.17/configuring.html> is a little brief in this regard:
Configuring Mime Types
TODO Mention non-standard paths, and custom mime type files
So I dared to ask whether you might encountered this and give me a hint were to look.
Probably the decorators, but I’ll find out

Best

Jochen


Von: David Pilato <notifications@github.com>
Antworten an: dadoonet/fscrawler <reply@reply.github.com>
Datum: Sonntag, 25. Februar 2018 um 19:10
An: dadoonet/fscrawler <fscrawler@noreply.github.com>
Cc: "Evertz, Jochen" <Jochen.Evertz@monotype.com>, Mention <mention@noreply.github.com>
Betreff: Re: [dadoonet/fscrawler] Enable custom tiki Parser (#498)


Do you perhaps know a starting point to do that when setting up a Tika parser from FSCrawler?

I'm confused. Isn't what you proposed with that:

{

  "name": "test",

  "fs": {

    "custom_tika_parsers": [

      {

        "class_name": "org.me.MyParser",

        "path_to_jar": "/some/full/path/to/myParser-0.0.1-SNAPSHOT.jar",

        "mime_types": ["application/dns", "or-another-mimetype-from-tika"]

      }

    ]

  }

}

I mean that you could create fscrawler-sample-parser.jar which you can add in a test?
If you don't know how to do it, no worries, I can do it in a follow up PR.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23issuecomment-2D368330762&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=59li_jqrR1bsrMi75LAHaIKTOiS0xEd6cuC3dA82x9o&s=0ARy8C33g04bMkqsMvRMT0UE2OuXW9_2x_wlW3_gSk4&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AUX8wtH-2DXHasZvMkLxsd-2DSJsaqsSR97sks5tYaIBgaJpZM4R3S2m&d=DwMFaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=59li_jqrR1bsrMi75LAHaIKTOiS0xEd6cuC3dA82x9o&s=_irdXd9ilfVn0aM61u9rpEw6B2pXGwg7BmFiwn-eXsU&e=>.


Jochen Evertz
Software Engineer
Monotype GmbH
Bergmannstr. 102
10961 Berlin
Germany
Phone:  +49 (30) 69596 221
Mobile: +49 (151) 17149034
                              
                              
monotype.com
Geschäftsführer / Managing Director: Christopher Kollat
Sitz Bad Homburg, HRB 10375, Amtsgericht Bad Homburg / Commercial Register of the Bad Homburg Local Court, HRB 10375

This e-mail message and any accompanying documentation is intended for the addressee only and may contain legally privileged and confidential information. If you are not the addressee any use of this information is strictly prohibited. If you have received this e-mail in error please notify us immediately by reply e-mail or telephone call on +49 (6172) 484-0 and delete this e-mail from your systems.

jevertz commented:
Sorry,I messed up the rebase.
I‘ll see to it to correct it.
We used perforce before, so I can only apologize for making noob git mistakes.

Von meinem iPhone gesendet



Jochen Evertz
Software Engineer
Monotype GmbH
Bergmannstr. 102
10961 Berlin
Germany
Phone:  +49 (30) 69596 221
Mobile: +49 (151) 17149034                                                            
monotype.com
Geschäftsführer / Managing Director: Christopher Kollat
Sitz Bad Homburg, HRB 10375, Amtsgericht Bad Homburg / Commercial Register of the Bad Homburg Local Court, HRB 10375
 
This e-mail message and any accompanying documentation is intended for the addressee only and may contain legally privileged and confidential information. If you are not the addressee any use of this information is strictly prohibited. If you have received this e-mail in error please notify us immediately by reply e-mail or telephone call on +49 (6172) 484-0 and delete this e-mail from your systems.
 
Am 01.03.2018 um 05:22 schrieb David Pilato <notifications@github.com<mailto:notifications@github.com>>:


Not sure what is happening with your branch but I can see a lot of changes which are not related to your PR.
May be you could rebase on master and while we are at it squash the changes?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dadoonet_fscrawler_pull_498-23issuecomment-2D369469940&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=3A6yFbWaB-_5uKRk0_iKOfjCAWACOENzFNEmFswak0M&s=7JgvvRnq2byZaWIgFCF3S8deM_qatmSn_4P09PE-0sE&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AUX8wo-2DlF6g2l8MFqplyA2XjwT6eylQMks5tZ3eBgaJpZM4R3S2m&d=DwMCaQ&c=euGZstcaTDllvimEN8b7jXrwqOf-v5A_CdpgnVfiiMM&r=rEdkK_qWWlll5TGtT1v3m_Yxzti17uKWa3FrzTh_rnE&m=3A6yFbWaB-_5uKRk0_iKOfjCAWACOENzFNEmFswak0M&s=BaY45xD3m9YtsCmBGrJRDsznDWfBv-seEGpR6c1IbxY&e=>.

dadoonet commented:
@jevertz Do you think you can come back with an updated PR at some point?

jevertz commented:
I'll see to do the rebase. Sorry for the delay, I was on holiday.
Probably It's best to include the changes manually again. To which version do you like me to base upon?

janhoy commented:
See #1004 for a related effort, which will let users take full control of what happens with a file before it gets indexed.
It could help solve this usecase, perhaps not as elegant by config but by giving you a hook to plug in your Java code.
